src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -341,56 +396,88 @@
-  (log/tracef "Deleting %d rows" (count pk-value-maps))
-  (with-jdbc-transaction [conn database-id]
-    (transduce
-      (m/indexed)
-      (fn
-        ([]
-         [[] []])
-        ([[errors]]
-         ;; if there were any errors throw an Exception with all the error messages.
-         (when (seq errors)
-           (.rollback conn)
-           (throw (ex-info (tru "Error(s) deleting rows.")
-                           {:status-code 400, :errors errors})))
-         ;; if there we no errors then return the created rows.
-         {:success true})
-        ([[errors] [row-index delete-filter]]
-         (try
-           (do-nested-transaction
-             driver
-             conn
-             (fn []
-               (actions/perform-action!*
-                 driver
-                 :row/delete
-                 database
-                 {:database   database-id
-                  :type       :query
-                  :query      {:source-table table-id
-                               :filter delete-filter}})))
-           [errors]
-           (catch Throwable e
-             [(conj errors {:index row-index, :error (ex-message e)})]))))
-      (pk-value-maps->honeysql-filter-clauses table-id pk-value-maps))))
+  [driver _action {database-id :id, :as database} {:keys [table-id rows]}]
+  (log/tracef "Deleting %d rows" (count rows))
+  (let [pk-name->id (table-id->pk-field-name->id table-id)]
+    ;; validate the keys in `rows`
+    (check-consistent-row-keys rows)
+    (check-rows-have-expected-columns-and-no-other-keys rows (keys pk-name->id))
+    (check-unique-rows rows)
+    ;; now do one `:row/delete` for each row
+    (perform-bulk-action-with-repeated-single-row-actions!
+     {:driver   driver
+      :database database
+      :action   :row/delete
+      :rows     rows
+      :xform    (comp (map (fn [row]
+                             {:database database-id
+                              :type     :query
+                              :query    {:source-table table-id
+                                         :filter       (row->mbql-filter-clause pk-name->id row)}}))
+                      #(completing % (fn [[errors _successes]]
+                                       (when (seq errors)
+                                         (throw (ex-info (tru "Error(s) deleting rows.")
+                                                         {:status-code 400, :errors errors})))
+                                       ;; `:bulk/delete` just returns a simple status message on success.
+                                       {:success true})))})))
+
+;;;; `bulk/update`
+
+(defn- bulk-update-row-xform
+  "Create a function to use to transform each row coming in to a `:bulk/update` request into an MBQL query that can be
+  passed to `:row/update`."
+  [{database-id :id, :as _database} table-id]
+  ;; TODO -- make sure all rows specify the PK columns
+  (let [pk-name->id (table-id->pk-field-name->id table-id)
```

Comment:
I tried to unify a couple of times but I ended up breaking things or making it more complicated. I'm not 100% sure it's actually better since the error conditions are different enough that trying to unify them makes them more complicated than leaving them separate.

A couple things are making refactoring this tricky:

- It turns out that there aren't tests for error responses in `bulk/delete` if you have all the PK columns **and** some non-PK columns. I accidentally broke this before I noticed
- It also seems like Postgres tests aren't running in CI. Some of the new bulk stuff fails in Postgres

I think we should try to figure that stuff out first and then maybe refactor this stuff if we can do it nicely